src/decoder.rs

```diff
@@ -546,117 +545,143 @@ pub enum ScopeEntry {
     Decoder(Decoder),
 }
 
-pub struct Scope<'a> {
-    parent: Option<&'a Scope<'a>>,
-    names: Vec<Cow<'static, str>>,
-    entries: Vec<ScopeEntry>,
+pub enum Scope<'a> {
+    Empty,
+    Normal(&'a NormalScope<'a>),
+    Single(SingleScope<'a>),
+    Decoder(DecoderScope<'a>),
 }
 
-pub struct ScopeIter<'a> {
-    parent: Option<&'a Scope<'a>>,
-    name_iter: std::iter::Rev<std::slice::Iter<'a, Cow<'static, str>>>,
-    entry_iter: std::iter::Rev<std::slice::Iter<'a, ScopeEntry>>,
+pub struct NormalScope<'a> {
+    parent: &'a Scope<'a>,
+    entries: Vec<(Cow<'static, str>, Value)>,
 }
 
-impl<'a> Iterator for ScopeIter<'a> {
-    type Item = (&'a Cow<'static, str>, &'a ScopeEntry);
-
-    fn next(&mut self) -> Option<Self::Item> {
-        match (self.name_iter.next(), self.entry_iter.next()) {
-            (Some(name), Some(entry)) => Some((name, entry)),
-            _ => match self.parent {
-                Some(parent) => {
-                    *self = parent.into_iter();
-                    self.next()
-                }
-                None => None,
-            },
-        }
-    }
+pub struct SingleScope<'a> {
+    parent: &'a Scope<'a>,
+    name: Cow<'static, str>,
+    value: &'a Value,
 }
 
-impl<'a> IntoIterator for &'a Scope<'a> {
-    type Item = (&'a Cow<'static, str>, &'a ScopeEntry);
+pub struct DecoderScope<'a> {
+    parent: &'a Scope<'a>,
+    name: Cow<'static, str>,
```

Comment:
Same idea as with `SingleScope<'a>`, because this is a short-lived stack allocation, we don't need a 'static/owned value, just a reference that lives at least as long as `'a`.